### PR TITLE
feat: support middleware processing chain

### PR DESCRIPTION
**Pull request by n063h**
refer to  #213 

- support user-implemented middlewares.
- move encode/decode functions into set_timeout context.
- add test case.

### CODE REVIEW

**Comment by VoVAllen** [comment]
What's the difference between middleware vs worker? Also in Python ecosystem, it's rare for framework to introduce the concept of middleware. I think people prefer decorator way for the AOP like behaviors

**Comment by kemingy** [comment]
> What's the difference between middleware vs worker? Also can we simplify the syntax using decorator?

Middleware should be part of the workers' forward process. More specifically, it should be `socket -> middleware -> forward -> middleware -> socket`.

Decorator and context manager also work for this purpose.

Back to this PR, I don't think users should use `partial` in their code.

Some thoughts:
- decorator should be easy to use if it's already implemented by us, but it may be hard to implement for users
- context manager is easier to implement compared to decorator
- both decorator and context manager can be nested

Welcome to more suggestions.

**Comment by VoVAllen** [comment]
```
    server.append_worker(
        Preprocess,
        timeout=100,
        middlewares=[
            partial(full_timer, "Preprocess"),
            DataLogger,
            partial(forward_timer, "Preprocess"),
        ],
    )
```

But here the syntax is per-worker. I think what you mean that middleware is per-server

**Comment by kemingy** [comment]
> But here the syntax is per-worker. I think what you mean that middleware is per-server

It should be per-worker.

Besides, I think we should reduce the complexity. The current implementation of plasma IPC wrapper is not elegant. It has some overlapping with the Mixin part. I guess users won't write their own IPC wrapper since the concept is already complex. We should refactor the IPC wrapper and this middleware design to make it easy to understand and use.

cc @lkevinzc

**Comment by n063h** [comment]
> ```
>     server.append_worker(
>         Preprocess,
>         timeout=100,
>         middlewares=[
>             partial(full_timer, "Preprocess"),
>             DataLogger,
>             partial(forward_timer, "Preprocess"),
>         ],
>     )
> ```
> 
> But here the syntax is per-worker. I think what you mean that middleware is per-server

Works of different stages could have shared/specific middlewares chain to process data. 
If per server,  it's ok but user shall deal with multi-stage conditions in one middleware class like `if self.stage==xx `, which can be less clear.

**Comment by n063h** [comment]
> Back to this PR, I don't think users should use `partial` in their code.

Using partial would be more flexible like
```python
    full_timer = partial(Timer, "full")
    forward_timer = partial(Timer, "forward")
```
If not using partial, user may have to implement FullTimer/ForwardTimer inherited from Timer?

**Comment by kemingy** [comment]
> > Back to this PR, I don't think users should use `partial` in their code.
> 
> Using partial would be more flexible like
> 
> ```python
>     full_timer = partial(Timer, "full")
>     forward_timer = partial(Timer, "forward")
> ```
> 
> If not using partial, user may have to implement FullTimer/ForwardTimer inherited from Timer?

I mean it should not exist at the interface level, no need to provide a hacking demo. It's okay to use it in the internal code when necessary.

**Comment by n063h** [comment]
The new commit's updates:
1. Plasma share-mem middleware test case
2. IPCMiddleware class, provide builtin stage-forward-strategy to simplify use.
    In `middleware_service` there's another way to implement PlasmaStore Middleware which could be more flexible, clear but complicated in the very begining, because user has to know the Coordinator pipeline.
3.  remove some uses of partial.

# Coordinator pipeline
for ipc middlewares like Plasma, Coordinator pipeline is like:

## Ingress Stage 
protocol_recv: data (list) -> deserialize data -> **No** PlasmaGet: data -> other middlewares/forward : data ->PlasmaPut: ids bytes -> serialize ids ->protocol_send

## Other Stage
protocol_recv: ids bytes -> deserialize ids: ids bytes -> PlasmaGet: data  -> other middlewares/forward : data -> PlasmaPut: ids bytes -> serialize ids ->protocol_send

## Egress Stage
protocol_recv: ids bytes -> deserialize ids: ids bytes -> PlasmaGet: data  -> other middlewares/forward : data -> **No** PlasmaPut: data -> serialize data ->protocol_send

File: mosec/middleware.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        and get data from another process in other stages.
+        User are out recommended to overwrite this method.
+
+        Arguments:
+            data: the input data
+
+        Returns:
+            the processed data, or the unprocessed data if the stage is STAGE_INGRESS.
+        """
+        if self.stage == STAGE_INGRESS:
```
Comment (kemingy) [inline mosec/middleware.py:197]: It can be both ingress and egress.

File: mosec/middleware.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    def _preprocess(self, ids: List[Any]) -> List[Any]:
+        """Process data before passing it to the next middleware or `forward`.
+
+        Arguments:
+            ids: the ID(s) of the data to be received.
+
+        Returns:
+            the processed data after it has been received.
+        """
+        return self.get(ids)
```
Comment (kemingy) [inline mosec/middleware.py:182]: What's the benefit of this abstraction?

**Review by kemingy — COMMENTED** [review]
The design of middleware looks similar to the worker.

Assume I was a user, if I have the code below:

```python
class Inf(Worker):
    def forward(self, data):
        self.pre(data)
        res = self.model.predict(data)
        self.post(data)
```

Then why should I change it to the current middleware style? What's the benefit?

The design has to be something people want to try.

**Comment by kemingy** [comment]
> The new commit's updates:
> 
>     1. Plasma share-mem middleware test case
> 
>     2. IPCMiddleware class, provide builtin stage-forward-strategy to simplify use.
>        In `middleware_service` there's another way to implement PlasmaStore Middleware which could be more flexible, clear but complicated in the very begining, because user has to know the Coordinator pipeline.

The plan is to refactor the whole IPC Wrapper. Plasma can be re-implemented with middleware or decorator or context manager or something else.

**Comment by n063h** [comment]
> Then why should I change it to the current middleware style? What's the benefit?

1. Reduce duplication of code. Some shared Middleware can be reused.
2. Better separation.  Split functionality into small middleware components, so in worker.forward users can focus on their computational business.

Welcome discuss on better design

**Comment by kemingy** [comment]
Closing this since it's implemented with mixin.

**Pull request closed without merging**